### ISSUE TITLE
[idea] tsconfig files based on file location.

### ISSUE THREAD

**Issue by trusktr** [TypeStrong/ts-node#655]
Before adding the `ignore` option, I saw that the concern was that it could fragment the ecosystem, and that there could be problem with different libs design to be compiled with different tsconfig settings, thus not a good idea.

However, compiling node modules works great internally for private modules, where someone or a team can guarantee that everyone is using the config. In this case things are just really really easy: just publish TypeScript files and done! With good caching, a `git pull` will only cause a few files to be re-built assuming caching works as expected. Easy as ice cream and pie!

So, here's a super duper awesome idea, although it might take some effort: it might be nice if `ts-node` could switch contexts inside of each node module, and if the module has it's own `tsconfig.json` file, the use that file to compile the files of that node module.

If done properly, this would solve both the ecosystem problem and clashing config problem.

I know, this is probably lots of work, but the result would be _super awesome_. Is it feasible?

(related issues for context: https://github.com/TypeStrong/ts-node/issues/158, https://github.com/TypeStrong/ts-node/issues/155)

**Comment by blakeembrey** [comment]
It might be possible, probably via upfront configuration instead of a default behaviour. Would that help you? One major hurdle we’ve run into is that TypeScript will refuse to compile inside node modules, so we will need to solve this too (I’m sure it’s just a configuration issue).

Another very challenging part is how this would all work together. For instance, some modules rely on also using the config file to do things such as override path resolution. This behaviour would probably be impossible to replicate all the way down at every level and may result in some inconsistencies when two different modules say differences on how paths or types are resolved (even if the types work, runtime may break since it’s acting differently at different levels). For these downsides I’d like to head a more thorough proposal and give myself time to think over any disadvantages of introducing this behaviour could bring.

Let’s start with this though, can you help explain why this would be preferable over having everything either compiled beforehand or transpiled by the same instance?

**Comment by trusktr** [comment]
> I’m sure it’s just a configuration issue

Yep, not problem compiling in node_modules here. In my tsconfig:

```js
    "include": [
        "src/**/*",
        "node_modules/some-package-that-i-want-to-compile/src/**/*"
    ]
```

> some modules rely on also using the config file to do things such as override path resolution

In my case, I remove that from ts-config, then use the [`module-alias`](https://www.npmjs.com/package/module-alias) package. It might be possible to use it programmatically in ts-node to call it based on what is in a project's tsconfig.

> can you help explain why this would be preferable over having everything either compiled beforehand or transpiled by the same instance?

The main reason being ease-of-use, just run ts-node in your project, and everything works; basically the end user never having to worry about node_modules compiling or not. But it's rather complicated on the ts-node implementation side.
